drivers/usb/device/usb_dc_smartbond.c

```diff
@@ -0,0 +1,1398 @@
+
+/* EP OUT buffers  */
+static uint8_t ep0_out_buf[EP0_OUT_BUF_SIZE];
+static uint8_t ep1_out_buf[EP1_OUT_BUF_SIZE];
+static uint8_t ep2_out_buf[EP2_OUT_BUF_SIZE];
+static uint8_t ep3_out_buf[EP3_OUT_BUF_SIZE];
+
+static uint8_t *const ep_out_bufs[4] = {
+	ep0_out_buf, ep1_out_buf, ep2_out_buf, ep3_out_buf,
+};
+
+static const uint16_t ep_out_buf_size[4] = {
+	EP0_OUT_BUF_SIZE, EP1_OUT_BUF_SIZE, EP2_OUT_BUF_SIZE, EP3_OUT_BUF_SIZE,
+};
+
+/* Node functional states */
+#define NFSR_NODE_RESET         0
+#define NFSR_NODE_RESUME        1
+#define NFSR_NODE_OPERATIONAL   2
+#define NFSR_NODE_SUSPEND       3
+/*
+ * Those two following states are added to allow going out of sleep mode
+ * using frame interrupt.  On remove wakeup RESUME state must be kept for
+ * at least 1ms. It is accomplished by using FRAME interrupt that goes
+ * through those two fake states before entering OPERATIONAL state.
+ */
+#define NFSR_NODE_WAKING        (0x10 | (NFSR_NODE_RESUME))
+#define NFSR_NODE_WAKING2       (0x20 | (NFSR_NODE_RESUME))
+
+struct smartbond_ep_reg_set {
+	volatile uint32_t epc_in;
+	volatile uint32_t txd;
+	volatile uint32_t txs;
+	volatile uint32_t txc;
+	volatile uint32_t epc_out;
+	volatile uint32_t rxd;
+	volatile uint32_t rxs;
+	volatile uint32_t rxc;
+};
+
+static struct smartbond_ep_reg_set *const reg_sets[4] = {
+	(struct smartbond_ep_reg_set *)&USB->USB_EPC0_REG,
+	(struct smartbond_ep_reg_set *)&USB->USB_EPC1_REG,
+	(struct smartbond_ep_reg_set *)&USB->USB_EPC3_REG,
+	(struct smartbond_ep_reg_set *)&USB->USB_EPC5_REG,
+};
+
+struct smartbond_ep_state {
+	atomic_t busy;
+	uint8_t *buffer;
+	uint16_t total_len;             /** Total length of current transfer */
+	uint16_t transferred;           /** Bytes transferred so far */
+	uint16_t mps;       /** Endpoint max packet size */
+	/** Packet size sent or received so far. It is used to modify transferred field
+	 * after ACK is received or when filling ISO endpoint with size larger then
+	 * FIFO size.
+	 */
+	uint16_t last_packet_size;
+
+	usb_dc_ep_callback cb;          /** Endpoint callback function */
+
+	uint8_t data1 : 1;              /** DATA0/1 toggle bit 1 DATA1 is expected or transmitted */
+	uint8_t stall : 1;              /** Endpoint is stalled */
+	uint8_t iso : 1;                /** ISO endpoint */
+	uint8_t ep_addr;                /** EP address */
+	struct smartbond_ep_reg_set *regs;
+};
+
+struct usb_dc_state {
+	bool vbus_present;
+	bool attached;
+	uint8_t nfsr;
+	usb_dc_status_callback status_cb;
+	struct smartbond_ep_state ep_state[2][4];
+	atomic_ptr_t dma_ep[2];         /** DMA used by channel */
+};
+
+static struct usb_dc_state dev_state;
+
+#define DA146XX_DMA_USB_MUX       (0x6 << ((DMA_CHANNEL_RX) * 2))
+#define DA146XX_DMA_USB_MUX_MASK  (0xF << ((DMA_CHANNEL_RX) * 2))
+
+struct smartbond_dma_channel {
+	uint32_t DMAx_A_START_REG;
+	uint32_t DMAx_B_START_REG;
+	uint32_t DMAx_INT_REG;
+	uint32_t DMAx_LEN_REG;
+	uint32_t DMAx_CTRL_REG;
+	uint32_t DMAx_IDX_REG;
```

Comment:
Variable and member identifier should always be lowercase, the only exception for USB support are member identifiers from the spec.